### ISSUE TITLE
CreateCustomToken() requires a certificate with "private_key"

### ISSUE THREAD

**Issue by frankspin89** [firebase/firebase-functions#3]
With the default configuration the private_key is not correctly set. Resulting in a error when calling createCustomToken():

```
Error: createCustomToken() requires a certificate with "private_key" set.
    at FirebaseAuthError.Error (native)
    at FirebaseAuthError.FirebaseError [as constructor] (/user_code/node_modules/firebase-admin/lib/utils/error.js:25:28)
    at new FirebaseAuthError (/user_code/node_modules/firebase-admin/lib/utils/error.js:90:23)
    at FirebaseTokenGenerator.createCustomToken (/user_code/node_modules/firebase-admin/lib/auth/token-generator.js:62:19)
    at Auth.createCustomToken (/user_code/node_modules/firebase-admin/lib/auth/auth.js:89:37)
    at /user_code/index.js:44:26
    at process._tickDomainCallback (internal/process/next_tick.js:129:7)
```

Changing the initializeApp function  to this will prevent the error:

```

const firebaseCredential = {
  "type": "service_account",
  "project_id": "",
  "private_key_id": "",
  "private_key": "",
  "client_email": "",
  "client_id": "",
  "auth_uri": "https://accounts.google.com/o/oauth2/auth",
  "token_uri": "https://accounts.google.com/o/oauth2/token",
  "auth_provider_x509_cert_url": "",
  "client_x509_cert_url": ""
};

admin.initializeApp({
  credential: admin.credential.cert(firebaseCredential),
  databaseURL: xxx.firebaseio.com'
});
```

Currently I have the firebaseCredential inline in my code. That's not ideal. Referencing a external .json file doesn't seems the work at the moment. And storing the firebaseCredentials in a environment config doesn't work either.

**Comment by jwngr** [comment]
In the future, please indicate when you cross post. I answered this one [Stack Overflow](http://stackoverflow.com/a/42724251/2955366). We hope to improve this situation in the future.

**Comment by mspado** [comment]
I don't think the issue should be closed. It's not fixed and is still breaking things.

It quietly broke a batch process when we switched to ```admin.initializeApp(functions.config().firebase);``` instead of a manual toggle.

**Comment by laurenzlong** [comment]
Please file in https://github.com/firebase/firebase-admin-node if you would like to see createCustomToken() supported with default credentials.

**Comment by manishtpatel** [comment]
reading 5.9.1 release note got me excited, thought this issue is resolved. Still not working with 5.9.1

[5.9.1 release note](https://github.com/firebase/firebase-admin-node/releases/tag/v5.9.1)
The admin.initializeApp() method can now be invoked without a credential option. This prompts the SDK to use Google Application Default Credentials.

What am i missing? What is this Google Application Default Credentials?

**Comment by laurenzlong** [comment]
Within the functions runtime, `admin.initializeApp(functions.config().firebase);` is still the right way to initialize an admin app, even with 5.9.1 release. And functions.config().firebase is already using Application Default Credentials (this is a credential that is available in a Google-managed server, such as that of Cloud Functions, so that you don't have to supply a service account key).

tl;dr: The 5.9.1 release doesn't change the situation, if you would like to see createCustomToken() supported with the default credentials already available in the functions runtime, please file a request in https://github.com/firebase/firebase-admin-node v

**Comment by sbnon** [comment]
`admin.initializeApp(functions.config().firebase)` appears to have been removed in Firebase Functions v1.0 (https://firebase.google.com/docs/functions/beta-v1-diff).

How do you now provide a custom certificate to Firebase with the v1.0 functions API?

**Comment by camstuart** [comment]
+1

The docs I'm reading say that if I am running in a cloud function all I need do is:
`admin.initializeApp(functions.config().firebase);`

But I get the same error about the private key missing.

I am following the docs here:
https://firebase.google.com/docs/admin/setup
